base_delivery_carrier_label/models/stock_picking.py

```diff
@@ -0,0 +1,253 @@
+    carrier_code = fields.Char(
+        related='carrier_id.code',
+        readonly=True,
+    )
+    option_ids = fields.Many2many(comodel_name='delivery.carrier.option',
+                                  string='Options')
+
+    def generate_default_label(self):
+        """ Abstract method
+
+        :return: (file_binary, file_type)
+
+        """
+        raise NotImplementedError(_('No label is configured for the '
+                                    'selected delivery method.'))
+
+    def generate_shipping_labels(self):
+        """Generate a shipping label by default
+
+        This method can be inherited to create specific shipping labels
+        a list of label must be return as we can have multiple
+        stock.quant.package for a single picking representing packs
+
+        :return: list of dict containing
+           name: name to give to the attachement
+           file: file as string
+           file_type: string of file type like 'PDF'
+           (optional)
+           tracking_number: tracking id defined by your carrier
+
+        """
+        self.ensure_one()
+        default_label = self.generate_default_label()
+        labels = []
+        for package in self._get_packages_from_picking():
+            pack_label = default_label.copy()
+            pack_label['tracking_number'] = package.id
+            labels.append(pack_label)
+        return labels
+
+    def get_shipping_label_values(self, label):
+        self.ensure_one()
+        return {
+            'name': label['name'],
+            'datas_fname': label.get('filename', label['name']),
+            'res_id': self.id,
+            'res_model': 'stock.picking',
+            'datas': base64.b64encode(label['file']),
+            'file_type': label['file_type'],
+        }
+
+    def generate_labels(self):
+        """ Legacy method. Remove me after 12.0
+        """
+        _logger.warning(
+            "Your delivery module depending on your carrier must call "
+            "action_generate_carrier_label() method "
+            "instead of generate_labels()")
+        return self.action_generate_carrier_label()
+
+    def _set_a_default_package(self):
+        """ Pickings using this module must have a package
+            If not this method put it one silently
+        """
+        for picking in self:
+            move_lines = picking.move_line_ids.filtered(
+                lambda s: not (s.package_id or s.result_package_id))
+            if move_lines:
+                package = self.env['stock.quant.package'].create({})
+                move_lines.write({'result_package_id': package.id})
+
+    def action_generate_carrier_label(self):
+        """ Method for the 'Generate Label' button.
+
+        It will generate the labels for all the packages of the picking.
+        Packages are mandatory in this case
+
+        """
+        for pick in self:
+            pick._set_a_default_package()
+            shipping_labels = pick.generate_shipping_labels()
```

Comment:
We could replace this part by something like : 
`shipping_labels = getattr(pick.carrier_id.carrier_type, _generate_shipping_labels)`

This way, in carrier submodules, we would just have to implement a method like : 
`mycarriertype_generate_shipping_labels` to implement the carrier specific code.

I see 2 advantages : 
1) I think it is more flexible, this way, it allow to easily override the method in a custom module for example, because with today's method, it seems complicated?
2) Also it seems cleaner and easier to read than to have a method with a big list of `if pick.carrier_id.carrier_type == ""...`

If just change slighty the code of submodules (you just have to rename the method) and I guess it would be acceptable now.
@yvaucher What do you think?

Reply:
Well, after second thought, it is not even so generic and the gain is really low, so, you can forget about it